src/executors/kernel_constructor.hpp

```diff
@@ -213,10 +215,11 @@ struct ExpressionTreeFunctor {
   SYCL_BLAS_INLINE ExpressionTreeFunctor(local_memory_t scratch,
                                          expression_tree_t t)
       : scratch_(scratch), t_(t) {}
-  SYCL_BLAS_INLINE void operator()(cl::sycl::nd_item<1> i) {
-    t_.adjust_access_displacement();
+  SYCL_BLAS_INLINE void operator()(cl::sycl::nd_item<1> i) const {
+    expression_tree_t& non_const_t=*const_cast<expression_tree_t*>(&t_);
+    non_const_t.adjust_access_displacement();
     ExpressionTreeEvaluator<using_local_memory, expression_tree_t,
-                            value_t>::eval(t_, scratch_, i);
+                            value_t>::eval(non_const_t, scratch_, i);
```

Comment:
I guess this is more of a long term question, but can the `eval` be changed to take a const object?

Reply:
Discussed internally

Reply:
The answer is complicated, but probably not :)